tuning/sft_trainer.py

```diff
@@ -115,13 +115,28 @@ def train(
         train_args.fsdp = ""
         train_args.fsdp_config = {"xla": False}
 
+    # Initialize the tracker early so we can calculate custom metrics like model_load_time.
+    tracker_name = train_args.tracker
+    if tracker_name == 'aim':
+        if tracker_config is not None:
+            tracker = AimStackTracker(tracker_config)
+        else:
+            logger.error("Tracker name is set to "+tracker_name+" but config is None.")
+    else:
+        logger.info('No tracker set so just set a dummy API which does nothing')
+        tracker = Tracker()
+
     task_type = "CAUSAL_LM"
+
+    model_load_time = time.time()
     model = AutoModelForCausalLM.from_pretrained(
         model_args.model_name_or_path,
         cache_dir=train_args.cache_dir,
         torch_dtype=get_torch_dtype(model_args.torch_dtype),
         use_flash_attention_2=model_args.use_flash_attn,
     )
+    model_load_time = time.time() - model_load_time
+    tracker.track(metric=model_load_time, name='model_load_time')
```

Comment:
@VassilisVassiliadis can you please explain what you had in mind for the metadata. I can certainly add a parameter like `extra_metadata` which can be passed to the tracker but I am trying to understand what can be covered other than the grouping via experiment name.

Reply:
The code that I linked annotates an experiment with key values which show up when you view the run on aim. 

You can imagine a dictionary that looks like this:

```json
{"gpu_model": "A100-80GB", "dataset_id": "my-amazing-dataset"}
```

This is nifty because you can query AIM to receive a list of all experiments that used a certain `gpu_model` or `dataset_id`, etc

If you prefer, I can open a PR for #34 after this one goes through so as not to bundle together too many changes.

Reply:
That is small enough Vassilis let me merge this in with this PR. Thanks for the clarification.

Reply:
do you suspect this extra metadata to be pased via main as arguments? as a file? or just to the train function.

Reply:
Personally I'd be satisfied with just the `aim_metadata` parameter in `train()` as I use fms-hf-tuning via its python API - i.e. directly invoke the train() method from inside my python code.

That said, users who run sft_trainer.py and have access to an AIM Server may want use something like `--aim_metadata` (or equivalent) to associate their run with metadata.

Reply:
@VassilisVassiliadis check now. Added a new API in tracker which can solve your problem. 

I have added a command line to sft trainer for extra metadata which accepts a loose json

Reply:
The one thing I'd change inside `train()` is guard accessing `tracker` via an `if tracker is not None:` statement just because the default value of tracker is None: https://github.com/foundation-model-stack/fms-hf-tuning/blob/5086918887b02c8ebd064aac9e19dc70625fe474/tuning/sft_trainer.py#L95

Reply:
Changed the default so we won't have to say `if tracker is not None` everywhere we track metrics.

Reply:
I think it'll be better to  better to keep the default value as `None` and have a check at the top of the function which instantiates the `tracker` object if its value is zero. This way the code doesn't create 2 tracker objects, one when python defines the `train()` method and another when `main()` calls `train()`.

Reply:
Done